File: Towers of Hanoi/hanoi.erl
-module(hanoi).  
-export([start/0]).  
  
start() ->  
    NDisks = 5,
    Source = {1, lists:seq(NDisks, 1, -1)},
    Destination = {3, []},
    Using = {2, []},
    Towers = [[Source]++[Using]++[Destination]],
    Moves = round(math:pow(2,NDisks))-1,
    io:format("~nTotal Steps for ~w pegs is ~w~n",[NDisks,Moves]),
    io:format("~nIterative moves : ~n"),
    IterativeResult = iterativeHanoi(NDisks, Towers),
    printHanoi(IterativeResult),
    io:format("~nRecursive moves : ~n"),
    [_,_,_,RecursiveResult] = recursiveHanoi(NDisks, Source, Destination, Using, Towers ),
    printHanoi(RecursiveResult),
    io:format("~nTailRecursive moves : ~n"),
    TailResult = tailRecursionHanoi(Moves+1, Towers),
    printHanoi(TailResult).

%% Recursive :
recursiveHanoi(1, {I1, Source}, {I2, Dest}, Using, Towers) -> 
    TempList = [{I1, lists:sublist(Source, length(Source) - 1)}] ++ [{I2, Dest++[lists:last(Source)]}] ++ [Using],
    SortList = lists:sort(TempList),
    [{I1, lists:sublist(Source, length(Source) - 1)}, {I2, Dest ++ [lists:last(Source)]}, Using, Towers ++ [SortList]];  
  
recursiveHanoi(NDisks, Source, Dest, Using, Towers) ->  
    [S1, U1, D1, T1] = recursiveHanoi(NDisks-1, Source, Using, Dest, Towers),  
    [S2, D2, U2, T2] = recursiveHanoi(1, S1, D1, U1, T1),  
    [U3, D3, S3, T3] = recursiveHanoi(NDisks-1, U2, D2, S2, T2),  
    [S3, D3, U3, T3].

%% Iteration :
iterativeHanoi(NDisks, Towers) ->  
    FinalList = lists:foldl(fun(Index, TempTowers) -> TempTowers ++ moveDisks(Index, lists:last(TempTowers)) end, Towers, lists:seq(1, round(math:pow(2, NDisks)) - 1)),
    FinalList.

%% move disks :
moveDisks(Index,PrevStep) ->
    From = (((Index band (Index - 1))) rem 3) + 1,
    To = (((Index bor (Index - 1)) + 1) rem 3) + 1,
    {I1, Lfrom}= lists:nth(From, PrevStep),
    {I2, Lto}= lists:nth(To, PrevStep),
    {I3, Ltemp}= lists:nth((6-To-From), PrevStep),
    TempList = [{I1,lists:sublist(Lfrom, length(Lfrom)-1)}] ++ [{I2, Lto++[lists:last(Lfrom)]}] ++ [{I3, Ltemp}],
    NextStep = lists:sort(TempList),
    [NextStep].

%% Tail Recursion :
tailRecursionHanoi(1, Towers) ->
    Towers;

tailRecursionHanoi(Steps, Towers) when Steps > 1 ->
    Index = length(Towers),
    NewList = Towers ++ moveDisks(Index,lists:last(Towers)),
    tailRecursionHanoi(Steps-1, NewList).


%% Printing the Lists :
printHanoi([]) ->
    ok;

printHanoi([First|Rest]) ->
    [S,T,D] = [V || {_,V} <- First],
    lists:foreach(fun(List) -> if length(List) == 0 -> io:format(" . "); true ->lists:foreach(fun(Value)-> io:format("~w", [Value]) end, List), io:format("  ")  end end, [S,T,D]),
    io:format("~n"),
    printHanoi(Rest).

File: Towers of Hanoi/Hanoi.scala
import scala.math.pow

object Hanoi {

  def printStands(lists: List[List[List[Int]]]): Unit =
  {
    for(value <- lists) {
      for (index <- value) {
        if (index.isEmpty) print(".") else index.foreach(print); print(" ")
      }
      println
    }
  }

  def moveDisk(list:List[List[List[Int]]], from:Int, to:Int): List[List[List[Int]]] = {
    val ln = list.length - 1;
    list ::: List(list(ln).updated(to, list(ln)(to) :+ (list(ln)(from).last)).updated(from, list(ln)(from).dropRight(1)))
  }

  def iterative(nDisks:Int, list:List[List[List[Int]]]): List[List[List[Int]]]  = {
    (1 to (pow(2, nDisks) - 1).toInt).foldLeft(list)((nList, I) =>moveDisk(nList, (I & I - 1) % 3, ((I | (I - 1)) + 1) % 3) )
  }

  def tail_recursion(nDisks:Int, list:List[List[List[Int]]], init:Int): List[List[List[Int]]]  = {
    val lists:List[List[List[Int]]] = list
    val x = init + 1 //# assume the next step
    val from = (x & x-1)% 3
    val to = ((x|(x-1))+1)%3
    val ln = lists.length - 1
    val nList = lists ::: List(lists(ln).updated(to, lists(ln)(to) :+ (lists(ln)(from).last)).updated(from, lists(ln)(from).dropRight(1)))
    if(nList(nList.length - 1).last.length >= nDisks)
      return nList
    tail_recursion(nDisks, nList, x)
  }

  def towerOfHanoiRecursive(nDisk:Int, bar1:Int, bar2:Int, bar3:Int, list:List[List[List[Int]]]):List[List[List[Int]]] =
  {
    val lists:List[List[List[Int]]] = list
    if(nDisk == 1) {
      val len = lists.length - 1
      lists ::: List(lists(len).updated(bar3, lists(len)(bar3) :+ nDisk).updated(bar1, lists(len)(bar1).dropRight(1)))
    }else {
      val nList = towerOfHanoiRecursive(nDisk - 1, bar1, bar3, bar2, lists)
      val nLen = nList.length - 1
      val sList = nList(nLen).updated(bar3, nList(nLen)(bar3) :+ nDisk).updated(bar1, nList(nLen)(bar1).dropRight(1))
      towerOfHanoiRecursive(nDisk - 1, bar2, bar1, bar3, nList ::: List(sList))
    }
  }

  def main(args:Array[String]): Unit =
  {
    val nDisks:Int = 5
    val lists = List(List((1 to nDisks).toList.reverse, List(), List()))
    println("Towers of Hanoi Recursive")
    printStands(towerOfHanoiRecursive(nDisks, 0, 1, 2, lists))
    println("Towers of Hanoi Iterative")
    printStands(iterative(nDisks, lists))
    println("Towers of Hanoi Tail Recursion")
    printStands(tail_recursion(nDisks, lists, 0))
  }
}